File: 0279-perfect-squares/0279-perfect-squares.cpp
class Solution {
public:
    int solve(int n,vector<int>& dp){
        if(n==0)return 0;
        if(dp[n]!=-1)return dp[n];
        int t1=INT_MAX;
        int j = 1,t=1;
        while(t<=n){
            t1 = min(t1,1 + solve(n-t,dp));
            j += 1;
            t = j*j; 
        }
        return dp[n] = t1;
    }
    int numSquares(int n) {
        vector<int> dp(n+1,-1);
        return solve(n,dp);
    }
};

File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [float('inf')]*(n+1)
        dp[0] = 0
        for i in range(1, n + 1):
            min_val = float('inf')

            j = 1
            while j * j <= i:
                min_val =min(min_val,dp[i - j * j] + 1)
                j += 1
            dp[i] = min_val
        
        return dp[n]